### ISSUE TITLE
`unused-void-attr` complains about unused voids that can't be removed

### ISSUE THREAD

**Issue by h1alexbel** [objectionary/lints#574]
Consider this EO snippet:

```eo
# Foo.
[] > foo
  boom > @
    filteredi.
      list
        * true false true
      v.not > [i] # <-- `i` is unused, but we cannot remove it
    * false
```

Here, `i` will be marked as unused void attribute by the `unused-void-attr` lint:

```text
The void attribute "i" is unused anywhere in the anonymous object (unused-void-attr (Single))
```

It is partially true, because, if we remove this `i` attribute, then we will get a runtime error:

```text
Can't put attribute with position 0 because it's not void one
```

Seems that such cases are false positives from `unused-void-attr`. How about we ignore void attributes in anonymous formations then?

**Comment by h1alexbel** [comment]
@yegor256 WDYT?

**Comment by yegor256** [comment]
@h1alexbel it's a valid concern, but the solution offered is not correct: it's not about anonymous formations, but any formations. Even if it's not anonymous, we may still have an attribute there that is unused, but must be present. The only possible solution I see is do what other language do (like Ruby for example): use certain naming pattern for such attributes. For example:

```
[a any1 b any2] > foo
  a.plus b > @
````

Here, `a` and `b` are the proper void attributes, while `any1` and `any2` are also void, but not expected to be used inside the object.

**Comment by h1alexbel** [comment]
@yegor256 maybe we also can employ such naming pattern like `_` or `/` to mark attributes that should be present and used inside, but not directly:

```eo
[a /any b /any2] > foo
  a.plus b > @
```

WDYT?

**Comment by yegor256** [comment]
@h1alexbel this will require changes in EO syntax, with only one motivation: to help static analysis. This motive is not strong enough for such a modification, IMHO.

**Comment by 0crat** [comment]
@h1alexbel Hey there, bug detective! 🕵️‍♂️ Great catch on that new bug - you've just scored yourself 12 points! Your sharp eyes are making our project better every day. If you spot anything else that [looks fishy](https://www.yegor256.com/2018/02/06/where-to-find-more-bugs.html) in the code, [don't be shy](https://www.yegor256.com/2014/04/13/bugs-are-welcome.html) about [flagging it](https://www.yegor256.com/2018/04/24/right-way-to-report-bugs.html). Keep it up, and watch your balance grow - you're already at [+317](https://www.eolang.org/zerocracy/objectionary-vitals.html) points!

**Comment by volodya-lombrozo** [comment]
@rultor release, tag is 0.2.5, title is 0.2.5

**Comment by volodya-lombrozo** [comment]
@rultor release, tag is `0.2.5`, title is `0.2.5`

**Comment by volodya-lombrozo** [comment]
@yegor256  While fixing this issue, we renamed void attributes with the `unused-` prefix as suggested by the lint. However, we discovered that this approach has a fundamental limitation.

**Finding:** It is impossible to statically determine whether a void attribute is truly unused, because any object — including one defined in a separate file — can access it by name after instantiation:

```
[x] > box   # x looks "unused" inside box's body

box 42 > b
b.x > @     # but x is perfectly accessible from outside
```

We verified this with a working `eoc` program. Renaming `x` to `unused-x` breaks `b.x` at runtime.

This means the `unused-void-attr` lint has false positives whenever a void attribute is part of the object's public interface -consumed by callers, child objects referencing parent scope, or Java runtime code setting attributes by name.

**Comment by yegor256** [comment]
@volodya-lombrozo I agree, let's get rid of this rule

**Comment by volodya-lombrozo** [comment]
@rultor release, tag is `0.2.6`, title is `0.2.6`

**Comment by volodya-lombrozo** [comment]
The rule was successfully removed
